### PR TITLE
HIVE-24867: Create iceberg-handler module in Hive

### PR DESCRIPTION
**Pull request by marton-bod**
1. da18e97 - this commit contains the below items. **This should not require any reviews.**
- creation of a new `iceberg-handler` module in Hive (with only skeleton pom)
- simple copy paste from the `iceberg-mr`: https://github.com/apache/iceberg/tree/master/mr/src into `hive/iceberg-handler/src`

2. efaeb74 - this commit contains the changes necessary to make the new iceberg-handler compile and pass its tests. **These are the changes that are _mostly_ non-trivial and should be reviewed.**
- Copy-pasted test classes from other iceberg modules which are used by our tests as well: 
-- `DeleteReadTests`, `FileHelpers`, `GenericAppenderHelper`, `RandomGenericData` from `iceberg-data`
-- `TestHiveMetastore` from `iceberg-hive-metastore`
-- `RandomUtil`, `AssertHelper`, `TestHelpers` from `iceberg-api`
The above are simple copy-pastes and **should not be reviewed,** except for:
-- `TestHelpers#assertSameSchemaList` commented out - 0.11.0 Schema does not contain schemaId() method
-- `TestHelpers#containsNaN` commented out - 0.11.0 does not contain this method yet
- Hive2 inspectors and their corresponding test classes replaced with their Hive3 counterparts where applicable. **These should not need to be reviewed as they are copy pastes**:
-- `IcebergDateObjectInspector` replaced with `IcebergDateObjectInspectorHive3`
-- `IcebergTimestampObjectInspector` replaced with `IcebergTimestampObjectInspectorHive3`
-- `IcebergTimestampWithZoneObjectInspector` replaced with `IcebergTimestampWithZoneObjectInspectorHive3`
- `HiveIcebergSerDe` and `TestHiveIcebergSerDe`: added new third parameter to the `initialize()` method, since the API changed in `AbstractSerDe.initialize()`
- `TestMetastore` change I: used the cdpd-master downstream version as the base for this file, which contains the changes related to `hiveWarehouseDir`/`hiveExternalWarehouseDir`
- `TestMetastore` change II: `HMSHandler` is not an inner class of `HiveMetaStore` anymore, but its own top-level class
- `TestMetastore` change III:  Uses `TestTxnDbUtils` to initialize the backend DB (see `start()` method). Removed the `setupMetastoreDB` as it's not necessary anymore.
- `TestIcebergObjectInspector`: removed the Hive2 cases for the date, timestamp and timestamptz cases
- `TestHiveShell`: added a few new test config settings
- `TestHiveIcebergStorageHandlerNoScan`: temporarily ignored `testIcebergAndHmsTableProperties` test (see comment)
- `TestHiveIcebergStorageHandlerNoScan`: temporarily ignored `testCreateTableWithColumnSpecificationHierarchy` test (see comment)
- Pom.xml dependencies should reflect exactly the dependencies declared in iceberg-mr build.gradle, except for:
-- Included `orc` as a test dependency because I got runtime errors otherwise
-- Did not include `jackson:annotations` because I got runtime errors otherwise
-- Hive-exec has been included without the core classifier, because I got runtime errors otherwise
-- Included `metastore-server` test jar, so that we access to `TestTxnDbUtils`
- Moved 6 test methods from `TestHiveIcebergStorageHandlerLocalScan` to `TestHiveIcebergStorageHandlerWithEngine`, because those tests now trigger engine executions
- `TestHiveIcebergStorageHandlerTimezone`: changed the engine from spark to tez, because these tests now also trigger engine executions
- `HiveIcebergFilterFactory#microsFromTimestamp`: using the cdpd-master downstream version here, same for the corresponding test in`TestHiveIcebergFilterFactory`
- Removed `mr` from the engine parameters in `TestHiveIcebergStorageHandlerWithEngine`
- Added `iceberg-handler` into `packaging` module

3. e78b7e2 - Changes iceberg-handler pom so that we build a shaded jar just like for druid-handler and kafka-handler (+ some small review changes)

### CODE REVIEW

File: iceberg-handler/pom.xml [hunk 1]
```diff
@@ -0,0 +1,189 @@
...
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>iceberg-handler</artifactId>
+    <packaging>jar</packaging>
+    <name>Hive Iceberg Handler</name>
+
+    <properties>
+        <hive.path.to.root>..</hive.path.to.root>
+        <iceberg-api.version>0.11.0</iceberg-api.version>
```
Comment (pvary) [inline iceberg-handler/pom.xml:19]: I would prefer to have this in the top level pom. `iceberg.version`
Why is this here?

File: iceberg-handler/pom.xml [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+                </exclusion>
+            </exclusions>
+        </dependency>
+
+        <dependency>
+            <groupId>org.apache.hive</groupId>
+            <artifactId>hive-exec</artifactId>
+            <version>${project.version}</version>
+            <exclusions>
+                <exclusion>
```
Comment (pvary) [inline iceberg-handler/pom.xml:73]: Why do we have these exclusions?

File: iceberg-handler/pom.xml [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+            </exclusions>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.hive</groupId>
+            <artifactId>hive-serde</artifactId>
+            <version>${project.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.hive</groupId>
+            <artifactId>hive-standalone-metastore-server</artifactId>
```
Comment (pvary) [inline iceberg-handler/pom.xml:118]: Is this a test depenency?

File: iceberg-handler/pom.xml (hunk 2 above)
Comment (marton-bod) [reply iceberg-handler/pom.xml:73]: The goal was to create the same dependency structure as much as possible as the original one: https://github.com/apache/iceberg/blob/master/build.gradle#L589

File: iceberg-handler/pom.xml (hunk 1 above)
Comment (marton-bod) [reply iceberg-handler/pom.xml:19]: Sure, will move it to the top level pom

File: iceberg-handler/pom.xml (hunk 3 above)
Comment (marton-bod) [reply iceberg-handler/pom.xml]: Yes, it's only a test dependency, so I've removed this one

File: iceberg-handler/pom.xml (hunk 1 above)
Comment (marton-bod) [reply iceberg-handler/pom.xml]: Moved it

File: iceberg-handler/pom.xml (hunk 2 above)
Comment (pvary) [reply iceberg-handler/pom.xml:78]: I would try to focus on keeping the same source files as in the Iceberg repo, so we can easily port changes between the 2, but otherwise I would not try to stick to the same things just because it was the same there.

OTOH if we know the reason why they were removed and it applies here too, then we should do the same

File: iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergFilterFactory.java [hunk 4]
```diff
@@ -190,6 +190,7 @@ private static int daysFromTimestamp(Timestamp timestamp) {
   // We have to use the LocalDateTime to get the micros. See the comment above.
   private static long microsFromTimestamp(Timestamp timestamp) {
     // `org.apache.hadoop.hive.common.type.Timestamp.valueOf(lit.toString()).toSqlTimestamp()`
+    // since HIVE-21862 changes literal parsing to UTC based timestamps
```
Comment (pvary) [inline iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergFilterFactory.java]: The comment in this form does not make sense to me 😢 
Either:
```
    // HIVE-21862 changes literal parsing to UTC based timestamps to this:
    // `org.apache.hadoop.hive.common.type.Timestamp.valueOf(lit.toString()).toSqlTimestamp()`
```
Or just remove the comment?

File: iceberg-handler/pom.xml [hunk 5]
```diff
@@ -16,41 +16,46 @@
 
     <properties>
         <hive.path.to.root>..</hive.path.to.root>
-        <iceberg-api.version>0.11.0</iceberg-api.version>
         <kryo-shaded.version>4.0.2</kryo-shaded.version>
-        <avro.iceberg.version>1.9.2</avro.iceberg.version>
+        <avro.iceberg.test.version>1.9.2</avro.iceberg.test.version>
```
Comment (pvary) [inline iceberg-handler/pom.xml:20]: Maybe it is just a little bit misleading name. We will use the 1.9.2 avro for Iceberg as well too.
Shall we just rename all the iceberg specific versions to:
- iceberg.kryo.version
- iceberg.avro.version
?

File: iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergFilterFactory.java [hunk 6]
```diff
@@ -230,7 +231,8 @@ public void testDateType() {
   public void testTimestampType() {
     Literal<Long> timestampLiteral = Literal.of("2012-10-02T05:16:17.123456").to(Types.TimestampType.withoutZone());
     long timestampMicros = timestampLiteral.value();
-    Timestamp ts = Timestamp.valueOf(DateTimeUtil.timestampFromMicros(timestampMicros));
+    // `org.apache.hadoop.hive.common.type.Timestamp.valueOf(lit.toString()).toSqlTimestamp()`
+    Timestamp ts = Timestamp.from(DateTimeUtil.timestampFromMicros(timestampMicros).toInstant(ZoneOffset.UTC));
```
Comment (pvary) [inline iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergFilterFactory.java:235]: Same problem than with the comment in `HiveIcebergFilterFactory`
Either remove the comment, or rephrase it

**Comment by pvary** [comment]
Checked the relevant files, left a few comments

File: iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergFilterFactory.java (hunk 6 above)
Comment (marton-bod) [reply iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergFilterFactory.java]: Removed it

File: iceberg-handler/pom.xml (hunk 5 above)
Comment (marton-bod) [reply iceberg-handler/pom.xml]: Good idea, done

File: iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergFilterFactory.java (hunk 4 above)
Comment (marton-bod) [reply iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergFilterFactory.java]: You're right, removed it

File: iceberg-handler/pom.xml (hunk 2 above)
Comment (marton-bod) [reply iceberg-handler/pom.xml:79]: Makes sense. As discussed offline, I've excluded orc, parquet, avro, guava and fasterxml only

**Comment by pvary** [comment]
Can we enable the Checkstyle and other code checks for the tests the same way as it is done in the Iceberg repo?

**Comment by marton-bod** [comment]
> Can we enable the Checkstyle and other code checks for the tests the same way as it is done in the Iceberg repo?
Sure, will do

**Comment by marton-bod** [comment]
@jcamachor Would you like to review as well?

File: iceberg-handler/pom.xml [hunk 7]
```diff
@@ -0,0 +1,252 @@
...
+         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <parent>
+        <groupId>org.apache.hive</groupId>
+        <artifactId>hive</artifactId>
+        <version>4.0.0-SNAPSHOT</version>
+        <relativePath>../pom.xml</relativePath>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <artifactId>iceberg-handler</artifactId>
```
Comment (szlta) [inline iceberg-handler/pom.xml:13]: Looks like the other handlers have this name pattern: hive-*-handler. Any reason the 'hive' part is omitted here?

Comment (marton-bod) [reply iceberg-handler/pom.xml:13]: Good question. I've seen two patterns: 

- druid, hbase and kudu had the hive- prefix 
- kafka-handler did not have the prefix

I went with the second one, but I'm fine with adding the prefix. I guess it's the majority :)

**Review by pvary — APPROVED** [review]
+1 pending tests
